### PR TITLE
(fleet) scaffolding of apt updater packaging

### PR DESCRIPTION
**Pull request by raphaelgavache**
### What does this PR do?

Add a simple build for an apt updater package. This is based on existing agent package.
Init scripts and systemd files are added following spec

Follow-ups
- tests that verify and enforces systemd behaviours
- tests on init scripts
- support rpm

### Test packages content
```
$ dpkg -e omnibus/pkg/datadog-updater_7.49.0~devel.git.1795.56d3c90-1_arm64.deb unpack_meta
tree unpack_meta
unpack_meta
├── control
├── md5sums
├── postinst
├── postrm
├── preinst
└── prerm
```
```
$ dpkg -x omnibus/pkg/datadog-updater_7.49.0~devel.git.1792.4bdbb3b-1_arm64.deb unpack_debug
❯ tree -l unpack_debug
unpack_debug
├── etc
│   └── datadog-agent
│       ├── conf.d
│       │   ├── apm.yaml.default
│       │   ├── container.d
│       │   │   └── conf.yaml.default
│       │   ├...
│       │   └── winproc.d
│       │       └── conf.yaml.default
│       └── datadog.yaml.example
├── lib
│   └── systemd
│       └── system
│           ├── datadog-agent-exp.service
│           ├── datadog-agent-process-exp.service
│           ├── datadog-agent-process.service
│           ├── datadog-agent-security-exp.service
│           ├── datadog-agent-security.service
│           ├── datadog-agent-sysprobe-exp.service
│           ├── datadog-agent-sysprobe.service
│           ├── datadog-agent-trace-exp.service
│           ├── datadog-agent-trace.service
│           ├── datadog-agent.service
│           ├── datadog-updater.service
│           ├── start-experiment.path
│           └── stop-experiment.path
├── opt
│   └── datadog
│       └── updater
│           ├── LICENSE
│           ├── LICENSES
│           │   ├── THIRD-PARTY-0BSD
│           │   ├── THIRD-PARTY-Apache-2.0
│           │   ├── THIRD-PARTY-BSD-2-Clause
│           │   ├── THIRD-PARTY-BSD-3-Clause
│           │   ├── THIRD-PARTY-ISC
│           │   ├── THIRD-PARTY-MIT
│           │   └── THIRD-PARTY-MPL-2.0
│           ├── bin
│           │   └── updater
│           │       └── updater
│           ├── embedded
│           │   ├── bin
│           │   └── lib
│           ├── run
│           ├── sources
│           │   ├── attr
│           │   │   └── offer.txt
│           │   ├── elfutils
│           │   │   └── offer.txt
│           │   ├── freetds
│           │   │   └── offer.txt
│           │   ├── gmp
│           │   │   └── offer.txt
│           │   ├── gnutls
│           │   │   └── offer.txt
│           │   ├── libacl
│           │   │   └── offer.txt
│           │   ├── libdb
│           │   │   └── offer.txt
│           │   ├── libgcrypt
│           │   │   └── offer.txt
│           │   ├── libgpg-error
│           │   │   └── offer.txt
│           │   ├── libsepol
│           │   │   └── offer.txt
│           │   ├── libtasn1
│           │   │   └── offer.txt
│           │   ├── libxcrypt
│           │   │   └── offer.txt
│           │   ├── m4
│           │   │   └── offer.txt
│           │   ├── nettle
│           │   │   └── offer.txt
│           │   ├── openscap
│           │   │   └── offer.txt
│           │   ├── procps-ng
│           │   │   └── procps-ng-procps-v3.3.16.tar.gz
│           │   ├── rpm
│           │   │   └── offer.txt
│           │   └── util-linux
│           │       └── offer.txt
│           ├── version-manifest.json
│           └── version-manifest.txt
└── var
    └── log
        └── datadog
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: omnibus/package-scripts/updater-deb/postinst
Comment (chouquette) [inline omnibus/package-scripts/updater-deb/postinst:75]: Don't we want to mention the install method is the updater?

File: omnibus/package-scripts/updater-deb/postrm
Comment (chouquette) [inline omnibus/package-scripts/updater-deb/postrm:18]: nit: tab in a file that uses spaces

File: omnibus/config/templates/updater/datadog-agent-exp.service.erb
Comment (chouquette) [inline omnibus/config/templates/updater/datadog-agent-exp.service.erb:8]: Could we use `Requires` here to avoid having to stop all dependent services manually when stopping the experiment?

File: tasks/updater.py
Comment (chouquette) [inline tasks/updater.py:25]: do we intent to support agent 6 and 7 through the updater? Or can we assume 7?

File: tasks/updater.py
Comment (chouquette) [inline tasks/updater.py:1]: The omnibus boilerplate could probably be imported and used directly from the existing agent script (ideally it could be moved to a common file but that's definitely out of scope here)

File: omnibus/config/templates/updater/datadog-agent-exp.service.erb
Comment (raphaelgavache) [reply omnibus/config/templates/updater/datadog-agent-exp.service.erb:17]: Yes, this is expected: if the agent exits with no error, it gets restarted 2 times and then systemd kills the experiment and runs back the stable agent

File: omnibus/package-scripts/updater-deb/postinst
Comment (bkabrda) [reply omnibus/package-scripts/updater-deb/postinst:75]: The install_info file will definitely need to change somehow. I would agree that this is good for initial testing, but we need to reflect the information here that the Agent is installed by the updater, but we also want to include the updater version and its install tool etc.

So I would tentatively accept this, but it will certainly need to be improved.

File: omnibus/config/templates/updater/datadog-agent-exp.service.erb
Comment (raphaelgavache) [reply omnibus/config/templates/updater/datadog-agent-exp.service.erb:8]: Good point, I'll test that change after when I add end to end tests on systemd

**Comment by bkabrda** [comment]
I think this looks good overall (obviously the points brought up by @chouquette should be addressed). I think the two bigger things that will need improvement are:

* The install_info file format will need to be amended to be able to handle information about the updater - alternatively we can have one install_info for Agent and then install_info_updater for the Updater. Either way the Agent will have to understand how to submit this new information and the backend code that parses it will need to be updated. I think we don't have to solve this right now, so I won't block the PR on it, but we *must* solve it before any real release.
* I'm thinking whether we need to stop/restart all the Agent services in the package scripts when we're updating the Updater. In theory, the Agent could continue running when the Updater package is *updating*, couldn't it?

File: tasks/updater.py
Comment (raphaelgavache) [reply tasks/updater.py]: only 7, I removed the parameter

File: omnibus/package-scripts/updater-deb/postinst
Comment (raphaelgavache) [reply omnibus/package-scripts/updater-deb/postinst:75]: I've added an extra tag `installer: updater`. I will dig deeper on how this is used

**Comment by raphaelgavache** [comment]
Good point, I've updated so the only interactions with the agent are
- disable + stop on uninstall
- enable the systemd files

Starting the agent is left to the updater and no restarts will happen when the updater is updated

**Review by bkabrda — APPROVED** [review]
Approving with a note that the install_info file modification is fine for now, but should be investigated again in more depth eventually.

**Pull request merged**
